When returning an unsized place from a closure, should suggest returning a reference to it instead

### Code

```Rust
fn main() {
    let o = Some("Hello, world!");
    for s in o.map(|s| s[3..8]) {}
}
```

### Current output

```Shell
error[E0277]: the size for values of type `str` cannot be known at compilation time
    --> src/main.rs:3:16
     |
   3 |     for s in o.map(|s| s[3..8]) {}
     |                ^^^ doesn't have a size known at compile-time
     |
     = help: the trait `Sized` is not implemented for `str`
note: required by an implicit `Sized` bound in `Option::<T>::map`
    --> /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:1160:22
     |
1160 |     pub const fn map<U, F>(self, f: F) -> Option<U>
     |                      ^ required by the implicit `Sized` requirement on this type parameter in `Option::<T>::map`

error[E0277]: the size for values of type `str` cannot be known at compilation time
 --> src/main.rs:3:24
  |
3 |     for s in o.map(|s| s[3..8]) {}
  |                        ^^^^^^^ doesn't have a size known at compile-time
  |
  = help: the trait `Sized` is not implemented for `str`
  = note: the return type of a function must have a statically known size

error[E0277]: the size for values of type `str` cannot be known at compilation time
   --> src/main.rs:3:14
    |
  3 |     for s in o.map(|s| s[3..8]) {}
    |              ^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
    |
    = help: the trait `Sized` is not implemented for `str`
note: required by an implicit `Sized` bound in `Option`
   --> /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:600:17
    |
600 | pub enum Option<T> {
    |                 ^ required by the implicit `Sized` requirement on this type parameter in `Option`

error[E0277]: `Option<str>` is not an iterator
    --> src/main.rs:3:14
     |
   3 |     for s in o.map(|s| s[3..8]) {}
     |              ^^^^^^^^^^^^^^^^^^ `Option<str>` is not an iterator
     |
     = help: the trait `IntoIterator` is not implemented for `Option<str>`
help: the following other types implement trait `IntoIterator`
    --> /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:2304:1
     |
2304 | impl<T> const IntoIterator for Option<T> {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `Option<T>`
...
2328 | impl<'a, T> IntoIterator for &'a Option<T> {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `&Option<T>`
...
2338 | impl<'a, T> IntoIterator for &'a mut Option<T> {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `&mut Option<T>`
```

### Desired output

```Shell
error[E0277]: the size for values of type `str` cannot be known at compilation time
 --> src/main.rs:3:24
  |
3 |     for s in o.map(|s| s[3..8]) {}
  |                        ^^^^^^^ doesn't have a size known at compile-time
  |
  = help: the trait `Sized` is not implemented for `str`
  = note: the return type of a function must have a statically known size
help: consider borrowing the return value
  |
3 |     for s in o.map(|s| &s[3..8]) {}
  |                        +
```

### Rationale and extra context

This error happens when the code attempts to return a place containing an unsized type from a closure (rather than a reference to that place). There are two things wrong with the resulting error messages.

One is that they assume throughout the rest of the compile that the return type (in this case, `str`) is correct (despite being impossible), which leads to a lot of secondary error messages about unsized types being used in a place where they can't be used: out of the four error messages here, only the second one (the one that lists the root cause, trying to return an unsized type) is actually useful. This can make it hard for a new user to figure out specifically where the problem is, as the other error messages are talking about side effects that the mistake has on the rest of the code rather than the mistake itself.

The other is that there is a likely fix (returning a reference to a place rather than trying to return the place itself) which should be suggested: in most cases, users who are trying to return a `str` or a slice from a function probably wanted to return a reference to it instead.

There may be cases where `&mut` is a better suggestion than `&`, but figuring that out would likely require a second type-check against a hypothetical type, so it may be better to just suggest `&` unconditionally and let the compiler suggest `&mut` on the next attempt.

The check should be specific to trying to return an unsized place that isn't a temporary – otherwise the user's mistake is likely to be different (and the suggestion of adding `&` wouldn't work anyway). It's also specific to closures with no explicitly stated return type (because when there is explicitly stated return type, either with a closure or a function, the error messages are fine already).

The code in this report is an MCVE reduced from an actual program created by a newcomer to Rust.


### Other cases

```Rust
// Effectively the same thing also happens with slices:
fn main() {
    let o = Some(b"Hello, world!");
    for s in o.map(|s| s[3..8]) {}
}
```

### Rust Version

```Shell
Tested on the Rust playground
Nightly version: 1.95.0-nightly (2026-02-02 f60a0f1bcc5a2a6dd8eb)
```

### Anything else?

@rustbot label +D-lack-of-suggestion +D-confusing +D-newcomer-roadblock +A-closures +A-DSTs +S-has-mcve

--- a/compiler/rustc_trait_selection/src/error_reporting/traits/fulfillment_errors.rs
+++ b/compiler/rustc_trait_selection/src/error_reporting/traits/fulfillment_errors.rs
@@ -2746,6 +2746,11 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
                 obligation.param_env,
                 obligation.cause.code(),
             );
+            self.suggest_borrow_for_unsized_closure_return(
+                obligation.cause.body_id,
+                err,
+                obligation.predicate,
+            );
             self.suggest_unsized_bound_if_applicable(err, obligation);
             if let Some(span) = err.span.primary_span()
                 && let Some(mut diag) =

--- a/compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs
+++ b/compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs
@@ -2193,6 +2193,60 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
         false
     }
 
+    pub(super) fn suggest_borrow_for_unsized_closure_return<G: EmissionGuarantee>(
+        &self,
+        body_id: LocalDefId,
+        err: &mut Diag<'_, G>,
+        predicate: ty::Predicate<'tcx>,
+    ) {
+        let Some(pred) = predicate.as_trait_clause() else {
+            return;
+        };
+        if !self.tcx.is_lang_item(pred.def_id(), LangItem::Sized) {
+            return;
+        }
+
+        let Some(span) = err.span.primary_span() else {
+            return;
+        };
+        let Some(node_body_id) = self.tcx.hir_node_by_def_id(body_id).body_id() else {
+            return;
+        };
+        let body = self.tcx.hir_body(node_body_id);
+        let mut expr_finder = FindExprBySpan::new(span, self.tcx);
+        expr_finder.visit_expr(body.value);
+        let Some(expr) = expr_finder.result else {
+            return;
+        };
+
+        let closure = match expr.kind {
+            hir::ExprKind::Call(_, args) => args.iter().find_map(|arg| match arg.kind {
+                hir::ExprKind::Closure(closure) => Some(closure),
+                _ => None,
+            }),
+            hir::ExprKind::MethodCall(_, _, args, _) => {
+                args.iter().find_map(|arg| match arg.kind {
+                    hir::ExprKind::Closure(closure) => Some(closure),
+                    _ => None,
+                })
+            }
+            _ => None,
+        };
+        let Some(closure) = closure else {
+            return;
+        };
+        if !matches!(closure.fn_decl.output, hir::FnRetTy::DefaultReturn(_)) {
+            return;
+        }
+
+        err.span_suggestion_verbose(
+            self.tcx.hir_body(closure.body).value.span.shrink_to_lo(),
+            "consider borrowing the value",
+            "&",
+            Applicability::MaybeIncorrect,
+        );
+    }
+
     pub(super) fn return_type_span(&self, obligation: &PredicateObligation<'tcx>) -> Option<Span> {
         let hir::Node::Item(hir::Item { kind: hir::ItemKind::Fn { sig, .. }, .. }) =
             self.tcx.hir_node_by_def_id(obligation.cause.body_id)

--- a/tests/ui/closures/unsized-return-suggest-ref-issue-152064.fixed
+++ b/tests/ui/closures/unsized-return-suggest-ref-issue-152064.fixed
@@ -1,0 +1,20 @@
+#![allow(unused_variables, for_loops_over_fallibles)]
+//@ run-rustfix
+
+fn main() {
+    // Basic case from the issue: str slice
+    let o = Some("Hello, world!");
+    for s in o.map(|s| &s[3..8]) {}
+    //~^ ERROR the size for values of type `str` cannot be known at compilation time
+    //~| ERROR the size for values of type `str` cannot be known at compilation time
+    //~| ERROR the size for values of type `str` cannot be known at compilation time
+    //~| ERROR `Option<str>` is not an iterator
+
+    // Byte slice case
+    let arr = Some(b"Hello, world!");
+    for s in arr.map(|s| &s[3..8]) {}
+    //~^ ERROR the size for values of type `[u8]` cannot be known at compilation time
+    //~| ERROR the size for values of type `[u8]` cannot be known at compilation time
+    //~| ERROR the size for values of type `[u8]` cannot be known at compilation time
+    //~| ERROR `Option<[u8]>` is not an iterator
+}

--- a/tests/ui/closures/unsized-return-suggest-ref-issue-152064.rs
+++ b/tests/ui/closures/unsized-return-suggest-ref-issue-152064.rs
@@ -1,0 +1,20 @@
+#![allow(unused_variables, for_loops_over_fallibles)]
+//@ run-rustfix
+
+fn main() {
+    // Basic case from the issue: str slice
+    let o = Some("Hello, world!");
+    for s in o.map(|s| s[3..8]) {}
+    //~^ ERROR the size for values of type `str` cannot be known at compilation time
+    //~| ERROR the size for values of type `str` cannot be known at compilation time
+    //~| ERROR the size for values of type `str` cannot be known at compilation time
+    //~| ERROR `Option<str>` is not an iterator
+
+    // Byte slice case
+    let arr = Some(b"Hello, world!");
+    for s in arr.map(|s| s[3..8]) {}
+    //~^ ERROR the size for values of type `[u8]` cannot be known at compilation time
+    //~| ERROR the size for values of type `[u8]` cannot be known at compilation time
+    //~| ERROR the size for values of type `[u8]` cannot be known at compilation time
+    //~| ERROR `Option<[u8]>` is not an iterator
+}

--- a/tests/ui/closures/unsized-return-suggest-ref-issue-152064.stderr
+++ b/tests/ui/closures/unsized-return-suggest-ref-issue-152064.stderr
@@ -1,0 +1,105 @@
+error[E0277]: the size for values of type `str` cannot be known at compilation time
+  --> $DIR/unsized-return-suggest-ref-issue-152064.rs:7:16
+   |
+LL |     for s in o.map(|s| s[3..8]) {}
+   |                ^^^ doesn't have a size known at compile-time
+   |
+   = help: the trait `Sized` is not implemented for `str`
+note: required by an implicit `Sized` bound in `Option::<T>::map`
+  --> $SRC_DIR/core/src/option.rs:LL:COL
+
+error[E0277]: the size for values of type `str` cannot be known at compilation time
+  --> $DIR/unsized-return-suggest-ref-issue-152064.rs:7:24
+   |
+LL |     for s in o.map(|s| s[3..8]) {}
+   |                        ^^^^^^^ doesn't have a size known at compile-time
+   |
+   = help: the trait `Sized` is not implemented for `str`
+   = note: the return type of a function must have a statically known size
+
+error[E0277]: the size for values of type `str` cannot be known at compilation time
+  --> $DIR/unsized-return-suggest-ref-issue-152064.rs:7:14
+   |
+LL |     for s in o.map(|s| s[3..8]) {}
+   |              ^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
+   |
+   = help: the trait `Sized` is not implemented for `str`
+note: required by an implicit `Sized` bound in `Option`
+  --> $SRC_DIR/core/src/option.rs:LL:COL
+help: consider borrowing the value
+   |
+LL |     for s in o.map(|s| &s[3..8]) {}
+   |                        +
+
+error[E0277]: `Option<str>` is not an iterator
+  --> $DIR/unsized-return-suggest-ref-issue-152064.rs:7:14
+   |
+LL |     for s in o.map(|s| s[3..8]) {}
+   |              ^^^^^^^^^^^^^^^^^^ `Option<str>` is not an iterator
+   |
+   = help: the trait `IntoIterator` is not implemented for `Option<str>`
+help: the following other types implement trait `IntoIterator`
+  --> $SRC_DIR/core/src/option.rs:LL:COL
+   |
+   = note: `Option<T>`
+  ::: $SRC_DIR/core/src/option.rs:LL:COL
+   |
+   = note: `&Option<T>`
+  ::: $SRC_DIR/core/src/option.rs:LL:COL
+   |
+   = note: `&mut Option<T>`
+
+error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
+  --> $DIR/unsized-return-suggest-ref-issue-152064.rs:15:18
+   |
+LL |     for s in arr.map(|s| s[3..8]) {}
+   |                  ^^^ doesn't have a size known at compile-time
+   |
+   = help: the trait `Sized` is not implemented for `[u8]`
+note: required by an implicit `Sized` bound in `Option::<T>::map`
+  --> $SRC_DIR/core/src/option.rs:LL:COL
+
+error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
+  --> $DIR/unsized-return-suggest-ref-issue-152064.rs:15:26
+   |
+LL |     for s in arr.map(|s| s[3..8]) {}
+   |                          ^^^^^^^ doesn't have a size known at compile-time
+   |
+   = help: the trait `Sized` is not implemented for `[u8]`
+   = note: the return type of a function must have a statically known size
+
+error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
+  --> $DIR/unsized-return-suggest-ref-issue-152064.rs:15:14
+   |
+LL |     for s in arr.map(|s| s[3..8]) {}
+   |              ^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
+   |
+   = help: the trait `Sized` is not implemented for `[u8]`
+note: required by an implicit `Sized` bound in `Option`
+  --> $SRC_DIR/core/src/option.rs:LL:COL
+help: consider borrowing the value
+   |
+LL |     for s in arr.map(|s| &s[3..8]) {}
+   |                          +
+
+error[E0277]: `Option<[u8]>` is not an iterator
+  --> $DIR/unsized-return-suggest-ref-issue-152064.rs:15:14
+   |
+LL |     for s in arr.map(|s| s[3..8]) {}
+   |              ^^^^^^^^^^^^^^^^^^^^ `Option<[u8]>` is not an iterator
+   |
+   = help: the trait `IntoIterator` is not implemented for `Option<[u8]>`
+help: the following other types implement trait `IntoIterator`
+  --> $SRC_DIR/core/src/option.rs:LL:COL
+   |
+   = note: `Option<T>`
+  ::: $SRC_DIR/core/src/option.rs:LL:COL
+   |
+   = note: `&Option<T>`
+  ::: $SRC_DIR/core/src/option.rs:LL:COL
+   |
+   = note: `&mut Option<T>`
+
+error: aborting due to 8 previous errors
+
+For more information about this error, try `rustc --explain E0277`.

